### PR TITLE
Deprecate attributes and expire deprecation in animation

### PR DESCRIPTION
**Pull request by oscargus**
## PR Summary

Not sure if the deprecation warning could actually be triggered without setting anim.save_count after creating the animation. Anyway, it's been around from 2.2 so I think this is OK.

The Python bug mentioned seems to have been fixed in 3.8(?): https://github.com/python/cpython/issues/74722

Privatized the save_count attribute as well as the `repeat` attribute. I cannot really see a use case for changing these after creating the animation.

## PR Checklist

**Tests and Styling**
- [N/A] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [N/A] New features are documented, with examples if plot related.
- [N/A] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [x] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

**Comment by oscargus** [comment]
I do not understand the warnings:
```
/home/circleci/project/doc/api/_as_gen/matplotlib.animation.ArtistAnimation.rst:33:<autosummary>:1: WARNING: py:obj reference target not found: matplotlib.animation.ArtistAnimation.repeat
/home/circleci/project/lib/matplotlib/animation.py:docstring of matplotlib.animation.FuncAnimation.new_frame_seq:1:<autosummary>:1: WARNING: py:obj reference target not found: matplotlib.animation.FuncAnimation.repeat
```

(I know that I've had similar issues with attributes earlier though...)

**Comment by QuLogic** [comment]
I don't think `save_count=None` was the intended removal, the default of 100 was; see #8278.

**Comment by oscargus** [comment]
Yeah, you are right. Forgot about the purpose a bit in...

**Comment by oscargus** [comment]
It seems like other parts of the code are relying on `save_count` being an integer (either provided or computed). So the behaviour discussed in #8278 is not straightforward to obain (which may be the reason it not has been removed since 2.2...). 

I guess that one can protect the line
```
self._save_seq = self._save_seq[-self._save_count:]
```
with a None-check though...

**Comment by QuLogic** [comment]
Then I think @anntzer has argued that it keeps running unless you specify `save_count`, just as it does with infinite generators elsewhere.

**Comment by oscargus** [comment]
> how do you save an infinitely running movie?

I guess that there is a file size limit:

```
Animation size has reached 21248815 bytes, exceeding the limit of 20971520.0. If you're sure you want a larger animation embedded, set the animation.embed_limit rc parameter to a larger value (in MB). This and further frames will be dropped.
```

**Comment by tacaswell** [comment]
@story645 is correct that there is an unbounded cache in the case of an inifinte generator that needs to be addressed.  The cap on the file size is also only in the jshtml export logic (so we do not make a multi-gigabyte html page 🤣, related if you want to see your browser break build our docs as `singlehtml`, but I digress).  

I pushed some commits that protect against this.  There might need to be some more thought around the default value of `cache_data_frames` as well.  I think this is going to hard break some use-cases, but those cases should also have been getting user visible warnings from 2.2 (which is 7 mpl versions).

Also working on the rebase and will add some tests.

**Comment by tacaswell** [comment]
and we are leaving the footcannon of trying to save an infinite generator, we just won't cache the data along the way.  I think we have to leave this as it is unlikely we will be able to solve the halting problem.

**Review by tacaswell — APPROVED** [review]
But I pushed a bunch of code to this PR so this should count as half.

File: doc/api/next_api_changes/behavior/24131-OG.rst [hunk 1]
```diff
@@ -0,0 +1,6 @@
+Passing None as ``save_count``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+... to `.FuncAnimation` no longer limits the number of frames to 100. Make
+sure that it either can be inferred from *frames* or provide an integer
+*save_count*.
```
Comment (timhoffm) [inline doc/api/next_api_changes/behavior/24131-OG.rst:6]:
```suggestion
``FuncAnimation(save_count=None)``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Passing ``save_count=None`` to `.FuncAnimation` no longer limits the number
of frames to 100. Make``` sure that it either can be inferred from *frames*
or provide an integer *save_count*.
```

FuncAnimation should be in the title to provide the context

File: lib/matplotlib/tests/test_animation.py [hunk 2]
```diff
@@ -330,9 +334,11 @@ def frames_generator():
 
             yield frame
 
+    MAX_FRAMES = 100
```
Comment (timhoffm) [inline lib/matplotlib/tests/test_animation.py:337]: Why not inline? 

```
save_count=100)
```

File: .circleci/config.yml [hunk 3]
```diff
@@ -222,6 +222,7 @@ jobs:
       - doc-deps-install
 
       - doc-build
+
```
Comment (timhoffm) [inline .circleci/config.yml:225]: Unrelated?

Comment (oscargus) [reply .circleci/config.yml:225]: Yes, this was a naïve attempt to see if it affected the execution of the next steps...

**Comment by oscargus** [comment]
I've tried to squash locally, but it just gets messed up. Please squash on merge...

**Comment by oscargus** [comment]
Feels like I've lost track of this, but I *think* that everything is considered now. The cleanliness may not be correct (in that case it is added yet again, since it is there).

So please squash upon merge... :-)

File: lib/matplotlib/animation.py [hunk 4]
```diff
@@ -1619,9 +1622,10 @@ def init_func() -> iterable_of_artists
         Whether frame data is cached.  Disabling cache might be helpful when
         frames contain large objects.
     """
+    save_count = _api.deprecate_privatize_attribute("3.7")
 
     def __init__(self, fig, func, frames=None, init_func=None, fargs=None,
-                 save_count=None, *, cache_frame_data=True, **kwargs):
+                 save_count=100, *, cache_frame_data=True, **kwargs):
```
Comment (oscargus) [inline lib/matplotlib/animation.py:1628]:
```suggestion
                 save_count=None, *, cache_frame_data=True, **kwargs):
```

**Comment by greglucas** [comment]
> I've tried to squash locally, but it just gets messed up. Please squash on merge...

You should be able to reset to a point you want to be at, including your remote, and then start cleaning up.

```bash
git reset --hard origin/animationdeprecations
git rebase -i upstream/main
# clean up
git push --force-with-lease
```

File: doc/api/next_api_changes/behavior/24131-OG.rst [hunk 5]
```diff
@@ -0,0 +1,6 @@
+``FuncAnimation(save_count=None)``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Passing ``save_count=None`` to `.FuncAnimation` no longer limits the number
+of frames to 100. Make``` sure that it either can be inferred from *frames*
```
Comment (QuLogic) [inline doc/api/next_api_changes/behavior/24131-OG.rst:5]:
```suggestion
of frames to 100. Make sure that it either can be inferred from *frames*
```

File: examples/animation/animate_decay.py [hunk 6]
```diff
@@ -50,5 +50,7 @@ def run(data):
 
     return line,
 
-ani = animation.FuncAnimation(fig, run, data_gen, interval=100, init_func=init)
+# Only save last 100 frames, but run forever
+ani = animation.FuncAnimation(fig, run, data_gen, interval=100, init_func=init,
+                              save_count=100)
```
Comment (QuLogic) [inline examples/animation/animate_decay.py:55]: Is this needed for sphinx-gallery? There isn't otherwise any saving going on in any of these examples.

Comment (tacaswell) [reply examples/animation/animate_decay.py:55]: yeah, SG saves both a gif thumbnail and a js_html output to embed in the docs.

**Comment by tacaswell** [comment]
I'm working on sorting out what is going on with the history here.

**Comment by tacaswell** [comment]
I think what happened is I had rebased and force-pushed, then Oscar merged that back into his branch and then added some more changes on top.

I started from just before the merge and then cherry-picked the exrta commits on top and force-pushed again.

**Comment by oscargus** [comment]
Yes, there was a bit of a hassle with commits, especially since I've been using two different computers. Have to learn how to deal with that in a better way...

(The main "issue" is that there will be a lot of intermediate conflicts while rebasing although the final diff will not have any conflicts...)

Just need to sort out the doc failure then. (And possibly use `privatize_attribute`? I seem to recall that it was used at some stage...)

**Comment by oscargus** [comment]
Tried to fix the doc-failure by updating missing references. Seems to make sense since this became a problem for a deprecated attribute that will go away soon anyway. (And it removes a few hundred lines from this file.)

**Review by greglucas — APPROVED** [review]

**Comment by tacaswell** [comment]
> (The main "issue" is that there will be a lot of intermediate conflicts while rebasing although the final diff will not have any conflicts...)

One trick I use (particuarly if I do not super care about the history or am willing to re-create it later) is to squash everything to one commit and then rebase.

File: lib/matplotlib/animation.py [hunk 7]
```diff
@@ -1652,19 +1653,31 @@ def iter_frames(frames=frames):
...
-            self.save_count = int(self.save_count)
+            self._save_count = frames
+            if save_count is not None:
+                _api.warn_external(
+                    f"You passed in an explicit {save_count=} which is being "
+                    f"ignored in favor of {frames=}."
+                )
+        if self._save_count is None and cache_frame_data:
+            _api.warn_external(
+                f"{frames=!r} which we can infer the length of, "
```
Comment (QuLogic) [inline lib/matplotlib/animation.py:1673]: Is this supposed to say "which we *cannot* infer the length of"?
